homeassistant/components/http/forwarded.py

```diff
@@ -0,0 +1,98 @@
+"""Middleware to handle forwarded data by a reverse proxy."""
+from ipaddress import ip_address
+import logging
+
+from aiohttp.hdrs import X_FORWARDED_FOR, X_FORWARDED_HOST, X_FORWARDED_PROTO
+from aiohttp.web import HTTPBadRequest, middleware
+
+from homeassistant.core import callback
+
+_LOGGER = logging.getLogger(__name__)
+
+# mypy: allow-untyped-defs
+
+
+@callback
+def setup_forwarded(app, use_x_forwarded_for, trusted_proxies):
+    """Create IP Ban middleware for the app."""
+
+    @middleware
+    async def forwarded_middleware(request, handler):
+        overrides = {}
+
+        # Handle X-Forwarded-For
+        forwarded_for = request.headers.getall(X_FORWARDED_FOR, [])
+        if not forwarded_for:
+            # No forwarding headers, continue as normal
+            return await handler(request)
+
+        # Multiple X-Forwarded-For headers
+        if len(forwarded_for) > 1:
+            _LOGGER.error("Too many headers for X-Forwarded-For", extra=request.headers)
+            raise HTTPBadRequest
+
+        # Process IP addresses in the forwarded for header
+        forwarded_for = list(reversed(forwarded_for[0].split(",")))
+        try:
+            forwarded_for = [
+                ip_address(addr) for addr in (a.strip() for a in forwarded_for) if addr
+            ]
+        except ValueError:
+            _LOGGER.error(
+                "Invalid IP address in X-Forwarded-For header", extra=request.headers
+            )
+            raise HTTPBadRequest
+
+        # Find the last trusted index in the X-Forwarded-For list
+        index = 0
+        for forwarded_ip in forwarded_for:
+            if any(forwarded_ip in trusted_proxy for trusted_proxy in trusted_proxies):
+                index += 1
+                continue
+            overrides["remote"] = str(forwarded_ip)
+            break
+
+        # If all the IP addresses are from trusted networks, take the left-most.
```

Comment:
Isn't that dangerous, now we are setting the IP to a known proxy?

Reply:
No, we are setting it to the client. The list is processed in reverse.

It starts on the last item first (actually the last item, since the list is reversed, closest by, the most trusted one), and starts stripped them out as long as they are in the trusted proxies list. The first untrusted one is considered the client IP.

However, if the request (first client in the list) originates from a trusted network, it was stripped out as well, hence: If the whole chain is trusted completely, this picks up the first item (client IP). As the list is reversed, this is the last.

Reply:
More interesting, should be log cases in which we could not verify the full proxy chain? (excluding the `client`, as that is expected to be untrusted).

Reply:
Hmm, maybe later, lets log about the connection source not trusted while a forwarded header was provided.